usr/src/lib/libzfs/common/libzfs_sendrecv.c

```diff
@@ -609,11 +621,23 @@ send_iterate_snap(zfs_handle_t *zhp, void *arg)
 {
 	send_data_t *sd = arg;
 	uint64_t guid = zhp->zfs_dmustats.dds_guid;
+	uint64_t txg = zhp->zfs_dmustats.dds_creation_txg;
 	char *snapname;
 	nvlist_t *nv;
 
 	snapname = strrchr(zhp->zfs_name, '@')+1;
 
+	if (sd->parent_tosnap_txg != 0 && txg > sd->parent_tosnap_txg) {
```

Comment:
Ah, I think I misunderstood what `parent_tosnap_txg` means.  From the name I thought it meant the txg of the tosnap of the parent filesystem.  E.g. in your example A/B's parent_tosnap_txg would be x1 (the parent of A/B is A, its tosnap is snap1, its txg is x1).  I think this member would be more accurately named `tosnap_txg` or `this_tosnap_txg`, because it is the txg of the tosnap of the filesystem we are currently examining (not its parent).